optax/_src/loss_test.py

```diff
@@ -238,5 +239,107 @@ def test_batched_predictions_only(self):
     np.testing.assert_allclose(out, self.exp_ys_only, atol=1e-5)
 
 
+def _lengths_to_paddings(lengths: chex.Array, maxlength: int) -> chex.Array:
+  indices = jnp.arange(maxlength).reshape((1,) * lengths.ndim + (maxlength,))
+  lengths = jnp.expand_dims(lengths, axis=-1)
+  elem_valid = indices < lengths
+  return np.logical_not(elem_valid).astype(np.float32)
+
+
+def _average_ctc_loss(logprobs: chex.Array, logprob_paddings: chex.Array,
+                      labels: chex.Array,
+                      label_paddings: chex.Array) -> chex.Array:
+  return jnp.average(
+      loss.ctc_loss(logprobs, logprob_paddings, labels,
+                    label_paddings))
+
+class CTCTest(parameterized.TestCase):
+  def setUp(self):
+    super().setUp()
+    np.random.seed(1234)
+
+  @chex.all_variants
+  def test_with_one_to_one_alignment(self):
+    # when inputsteps and outputsteps are equal, no blank will be allowed.
+    batchsize = 8
+    steps = 50
+    nclasses = 40
+    logits = np.random.randn(batchsize, steps, nclasses)
+    labels = np.random.uniform(
+        1, nclasses, size=(batchsize, steps)).astype(np.int32)
+
+    # This function only covers the cases without same-label repetition.
+    # `test_repeat_with_one_to_one_alignment` below complements those cases.
+    # So, redraw the samples for satisfying the non-repetition constraint.
+    for n in range(labels.shape[0]):
+      for t in range(1, labels.shape[1]):
+        while labels[n, t] == labels[n, t - 1]:
+          labels[n, t] = np.random.uniform(1, nclasses)
+
+    per_seq_loss = self.variant(loss.ctc_loss)(
+        logits, np.zeros(logits.shape[:2]), labels, np.zeros(labels.shape))
+
+    logprobs = jax.nn.log_softmax(logits)
+    for b in range(batchsize):
+      p = 0.0
+      for t in range(steps):
+        p += logprobs[b, t, labels[b, t]]
+      np.testing.assert_allclose(jnp.array(-p), per_seq_loss[b],
+                                 rtol=1e-6)
+
+  @chex.all_variants
+  def test_with_one_to_one_alignment_and_paddings(self):
+    # test if padding indicators work as expected.
+    batch_size = 1
+    nclasses = 8
+    steps = 4
+    logits = np.random.normal(size=[batch_size, steps, nclasses])
+    logprobs = jax.nn.log_softmax(logits)
+    labels = np.array([[1, 2, 3, 4]])
+    logit_padding = np.array([[0, 0, 0, 1]], dtype=np.float32)
+    label_padding = np.array([[0, 0, 0, 1]], dtype=np.float32)
+
+    value_and_grad = self.variant(jax.value_and_grad(_average_ctc_loss))
+    actual_loss, actual_gradients = value_and_grad(
+        logits, logit_padding, labels, label_padding)
+    expected_loss = -(logprobs[0, 0, 1] + logprobs[0, 1, 2] + logprobs[0, 2, 3])
+    np.testing.assert_allclose(expected_loss, actual_loss)
+
+    expected_gradients = np.array(jax.nn.softmax(logits))
+    expected_gradients[0, 0, 1] -= 1.0
```

Comment:
can we compute the gradients with jax (instead of manually computing them)

Reply:
The current gradient computation is already depending on Jax's grad (no custom backward implemented).
So, I think it's more important to compare it with manually-computed values.